### PR TITLE
Changes in common from the SID branch

### PR DESCRIPTION
**Pull request by anstaf**
- New stuff:
  - `tuple_util`:  `concat`, `concat_ex` and `flatten_ex`
  - `hymap`:  `concat`, `merge`
- Bug fixing and minor modifications in 'common' and 'meta' 
- change interface of `ij_cache`, `shared_allocator`, `tmp_storage_sid`
- introduce `launch_kernel`
- `positional` now is templated by the dimension and holds a single integer instead of triple.
- `sid/allocator` library is introduced
- composite variations are added  to `sid::multi_shift` and `sid::get_stride`
- `GT_CONSTEXPR` defined to nothing for nvcc
- the usage pg `GT_CONSTEXPR` and `wstd::move`/`wstd::forward` has been reduced to the `tuple_util` and `hymap` libraries and the stuff that depends on them.

### CODE REVIEW

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
do perftest

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
launch icgrid

**Comment by anstaf** [comment]
launch strgrid

**Comment by anstaf** [comment]
do perftest

File: unit_tests/stencil_composition/sid/test_sid_concept.cpp [hunk 1]
```diff
@@ -105,7 +105,7 @@ namespace gridtools {
                 std::is_same<std::decay_t<decltype(sid::get_origin(std::declval<testee &>())())>, testee *>(), "");
             static_assert(std::is_same<decltype(sid::get_strides(testee{})), strides>(), "");
 
-            GT_CONSTEXPR auto stride = sid::get_stride<void>(strides{});
+            auto stride = sid::get_stride<void>(strides{});
             static_assert(stride == 0, "");
```
Comment (lukasm91) [inline unit_tests/stencil_composition/sid/test_sid_concept.cpp]: how does this static_assert work, if it is not constexpr?

File: unit_tests/common/test_tuple_util.cpp [hunk 2]
```diff
@@ -214,6 +214,14 @@ namespace gridtools {
             EXPECT_EQ(90, acc);
         }
 
+        TEST(concat, functional) {
+            EXPECT_EQ(concat(std::make_tuple(1, 2), std::make_tuple(3, 4)), std::make_tuple(1, 2, 3, 4));
```
Comment (lukasm91) [inline unit_tests/common/test_tuple_util.cpp]: Is that enough, only one test? Thinking about corner cases (empty tuple or so).

File: include/gridtools/storage/common/storage_info.hpp [hunk 3]
```diff
@@ -237,23 +238,23 @@ namespace gridtools {
...
+        GT_FUNCTION constexpr uint_t padded_length() const {
             return m_padded_lengths[Dim];
         }
 
         /**
          * @brief Returns the array of padded_lengths, the lengths including the halo points (the outer region) and
          * padding.
          */
-        GT_FUNCTION GT_CONSTEXPR const array<uint_t, ndims> &padded_lengths() const { return m_padded_lengths; }
+        GT_FUNCTION constexpr const_ref<array<uint_t, ndims>> padded_lengths() const { return m_padded_lengths; }
```
Comment (lukasm91) [inline include/gridtools/storage/common/storage_info.hpp]: if you use `const_ref` for `padded_lengths`, we should probably also use it for `total_lengths` (at least if they are used in constexpr context).

File: include/gridtools/stencil_composition/sid/sid_shift_origin.hpp [hunk 4]
```diff
@@ -50,9 +50,11 @@ namespace gridtools {
                 friend UpperBounds const &sid_get_upper_bounds(shifted_sid const &obj) { return obj.m_upper_bounds; }
 
               public:
+                shifted_sid() = default;
```
Comment (lukasm91) [inline include/gridtools/stencil_composition/sid/sid_shift_origin.hpp]: what's a default constructed `shifted_sid`?

File: include/gridtools/stencil_composition/backend_cuda/launch_kernel.hpp [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            uint_t yblocks = (j_size + BlockSizeJ - 1) / BlockSizeJ;
+
+            dim3 blocks = {xblocks, yblocks, zblocks};
+            dim3 threads = {BlockSizeI, BlockSizeJ, 1};
+
+            if (i_size % BlockSizeI == 0 && j_size % BlockSizeJ == 0)
+                launch_kernel_impl_::trivial_wrapper<num_threads>
+                    <<<blocks, threads, shared_memory_size>>>(std::move(fun));
+            else
+                launch_kernel_impl_::zero_extent_wrapper<num_threads, BlockSizeI, BlockSizeJ>
```
Comment (lukasm91) [inline include/gridtools/stencil_composition/backend_cuda/launch_kernel.hpp]: do we gain a lot with it? has it compile time impact as we now compile to kernel launches?

File: include/gridtools/stencil_composition/sid/allocator.hpp [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+ *
+ *
+ *  API
+ *  ---
+ *
+ *  The library provides two types that model the concept:
+ *    - `allocator`,
+ *    - `cached_allocator`.
+ *
+ *  Both are templated with the functor that takes the size in bytes and returns std::unique_ptr
```
Comment (lukasm91) [inline include/gridtools/stencil_composition/sid/allocator.hpp]:
```suggestion
 *  Both are templated with the functor that takes the size in bytes and returns `std::unique_ptr`
```

File: include/gridtools/common/defs.hpp [hunk 7]
```diff
@@ -22,7 +22,11 @@
    @brief global definitions
 */
 
+#ifdef __CUDACC__
+#define GT_CONSTEXPR
```
Comment (lukasm91) [inline include/gridtools/common/defs.hpp]: If we do this with `constexpr` / `wstd` I would prefer this being in a separate PR because obviously that is a likely candidate for performance regressions...

File: include/gridtools/stencil_composition/sid/allocator.hpp (hunk 6 above)
Comment (anstaf) [reply include/gridtools/stencil_composition/sid/allocator.hpp]: done

File: include/gridtools/stencil_composition/sid/sid_shift_origin.hpp (hunk 4 above)
Comment (anstaf) [reply include/gridtools/stencil_composition/sid/sid_shift_origin.hpp]: removed

File: include/gridtools/common/hymap.hpp [hunk 8]
```diff
@@ -30,13 +30,20 @@
...
  *  The Default Behaviour
  *  ---------------------
  *  If hymap doesn't provide `hymap_get_keys`, the default is taken which is:
  *  `meta::list<integral_constant<int, 0>, ..., integral_constant<int, N> >` where N is `tuple_util::size` of hymap.
  *  This means that any `tuple_like` automatically models Hymap. And for the plain `tuple_like`'s you can use
  *  `at_key<integral_constant<int, N>>(obj)` instead of `tuple_util::get<N>(obj)`.
  *
+ *  There are two default `hymap_from_keys_values`: is provided if for hymaps that are structured like:
+ *     `SomeKeyTemplate<key_a, key_b, key_c>::values<val_a, val_b, val_c>`;
+ *     It just use `SomeKeyTemplate` with nested `values` to create a type.
```
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/common/hymap.hpp]: done

File: include/gridtools/stencil_composition/arg.hpp [hunk 9]
```diff
@@ -116,7 +116,9 @@ namespace gridtools {
 
         // replace the storage_info ID contained in a given storage with the new value
         template <unsigned Id, typename T>
-        struct tmp_data_store;
+        struct tmp_data_store {
```
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/stencil_composition/arg.hpp]: This serves the case where the type provided with the temporary arg is not gridtools::data_store. Used in the test for cuda temporaries. It is a transient solution. After the full refactoring the logic below will gone (assigning storage info id for the temporaries).

File: include/gridtools/common/gt_math.hpp [hunk 10]
```diff
@@ -11,6 +11,7 @@
 #include <cmath>
 
 #include "defs.hpp"
+#include "generic_metafunctions/const_ref.hpp"
```
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/common/gt_math.hpp]: removed

File: include/gridtools/common/boollist.hpp [hunk 11]
```diff
@@ -12,6 +12,7 @@
 #include "../meta/type_traits.hpp"
 #include "array.hpp"
 #include "defs.hpp"
+#include "generic_metafunctions/const_ref.hpp"
```
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/common/boollist.hpp]: removed

File: include/gridtools/common/cuda_util.hpp [hunk 12]
```diff
@@ -45,7 +45,7 @@ namespace gridtools {
         template <class T>
         unique_cuda_ptr<T> cuda_malloc(size_t size = 1) {
             T *ptr;
-            GT_CUDA_CHECK(cudaMalloc(&ptr, size * sizeof(T)));
+            GT_CUDA_CHECK(cudaMalloc(&ptr, size * sizeof(std::conditional_t<std::is_void<T>::value, char, T>)));
```
[the remark this answers is not in the document]
Comment (anstaf) [reply include/gridtools/common/cuda_util.hpp]: removed

File: unit_tests/stencil_composition/sid/test_sid_concept.cpp (hunk 1 above)
Comment (anstaf) [reply unit_tests/stencil_composition/sid/test_sid_concept.cpp]: it works, probably by accident. changed anyway.

File: include/gridtools/storage/common/storage_info.hpp (hunk 3 above)
Comment (anstaf) [reply include/gridtools/storage/common/storage_info.hpp]: done

File: unit_tests/common/test_tuple_util.cpp (hunk 2 above)
Comment (anstaf) [reply unit_tests/common/test_tuple_util.cpp]: Added one more. Ready to add more on request.

File: include/gridtools/stencil_composition/backend_cuda/launch_kernel.hpp (hunk 5 above)
Comment (anstaf) [reply include/gridtools/stencil_composition/backend_cuda/launch_kernel.hpp]: yes, it's a trade off. I have added TODO to investigate if it has positive impact.

**Comment by anstaf** [comment]
launch strgrid

**Comment by havogt** [comment]
launch perftest
